middleware/ratelimiter.go

```diff
@@ -0,0 +1,511 @@
+package middleware
+
+import (
+	"errors"
+	"github.com/labstack/echo/v4"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+	"fmt"
+	"strings"
+)
+type (
+	// RateLimiterConfig defines the config for RateLimiter middleware.
+	RateLimiterConfig struct {
+		// Skipper defines a function to skip middleware.
+		Skipper Skipper
+
+		LimitConfig LimiterConfig
+		//key prefix, default is "LIMIT:".
+		Prefix   string
+
+		// Use a redis client for limiter, if omit, it will use a memory limiter.
+		Client   RedisClient
+
+		//If request gets a  internal limiter error, just skip the limiter and let it go to next middleware
+		SkipRateLimiterInternalError	bool
+
+		//TODO: WhiteList
+	}
+	// LimiterConfig defines the limitation of middleware
+	LimiterConfig struct {
+		//The max count in duration for no policy, default is 100.
+		Max      int
+		//ip or header(key)
+		Strategy string
+		//Count duration for no policy, default is 1 Minute.
+		Duration time.Duration
+		//If the strategy is header which header key will be used for limits. such as Header("Client-Token")
+		Key           string
+	}
+
+	limiter struct {
+		abstractLimiter
+		prefix string
+	}
+
+	// Result of limiter.Get
+	Result struct {
+		Total     int           // It Equals Options.Max, or policy max
+		Remaining int           // It will always >= -1
+		Duration  time.Duration // It Equals Options.Duration, or policy duration
+		Reset     time.Time     // The limit record reset time
+	}
+
+	abstractLimiter interface {
+		getLimit(key string, policy ...int) ([]interface{}, error)
+		removeLimit(key string) error
+	}
+
+	//RedisClient interface
+	RedisClient interface {
+		DeleteKey(string) error
+		EvalulateSha(string, []string, ...interface{}) (interface{}, error)
+		LuaScriptLoad(string) (string, error)
+	}
+
+)
+
+var (
+	// DefaultRateLimiterConfig is the default rate limit middleware config.
+	DefaultRateLimiterConfig = RateLimiterConfig{
+		Skipper:      DefaultSkipper,
+		LimitConfig:LimiterConfig{
+			Max:100,
+			Duration: time.Minute * 1,
+			Strategy:"ip",
+			Key:"",
+		},
+		Prefix:"LIMIT",
+		Client:nil,
+		SkipRateLimiterInternalError:false,
+	}
+	limiterImp *limiter
```

Comment:
To support the use-case of creating multiple rate limiters (1 for minute, hour, day, etc...) can we move this from global to instance based?

Reply:
imo it makes thinks complex. What you have in your mind, can you give more specific example?

Reply:
this is a way to support multiple rate limiters. 
```
rateLimiter1:=middleware.RateLimiterWithConfig(middleware.RateLimiterConfig{
		LimitConfig: middleware.LimiterConfig{
			Max:                         2,
			Strategy:                    "header",
			Duration:                    time.Minute,
			Key:                         "test",
			HeaderTokenExtractorHandler: func(headerValue string) string {
				return headerValue
			},
		},
	})
	rateLimiter2:=middleware.RateLimiterWithConfig(middleware.RateLimiterConfig{
		LimitConfig: middleware.LimiterConfig{
			Max:                         2,
			Strategy:                    "ip",
			Duration:                    time.Minute,
		},})
	
	e.GET("/", func(c echo.Context) error {
		return c.String(http.StatusOK, "Hello, World!")
	},rateLimiter1)

	e.GET("/status", func(c echo.Context) error {
		return c.String(http.StatusOK, "Hello, World!")
	},rateLimiter2)
```

Reply:
Are you sure? I think in your example here, `rateLimiter2`'s config would overwrite `rateLimiter1`, so yes, you'd have 2 rate limiters, but both would be using the same `limiterImp`

Reply:
>can you give more specific example

The most specific example I can think of is:

```golang
e.GET("/", func(c echo.Context) error {
    return c.String(http.StatusOK, "Hello, World!")
}, rateLimiter1Min, rateLimiter60Min, rateLimiter1440Min)
```

For example, we could set 10,000 requests max per day, but we don't want all 10,000 to happen in the first minute, so the minute rate limiter could be 60.

Reply:
What I've presented above, is simple/clean, and I've seen in some node.js libraries, but there's another approach that's even more interesting.

Dynamic rate limits per request.

For example:

```golang
e.GET("/", func(c echo.Context) error {
    return c.String(http.StatusOK, "Hello, World!")
}, createRateLimiterMiddleware(func(c *echo.Context) ([]middleware.RateLimiterConfig, error) {
  user := extractUse(c)
  path := extractPath(c)

  if user.role == "premium" {
    return []middleware.RateLimiterConfig{middleware.RateLimiterConfig{Max: 1000, Duration: time.Hour}}, nil
  }

  return defaultRateLimiterConfigs(), nil
}))
```

Reply:
> Are you sure? I think in your example here, `rateLimiter2`'s config would overwrite `rateLimiter1`, so yes, you'd have 2 rate limiters, but both would be using the same `limiterImp`

you're right. there is only one limiterimp so first registered middleware is working. as you said it should work with many limiters.